### ISSUE TITLE
Related `and/or` conditions

### ISSUE THREAD

**Issue by steve-chavez** [PostgREST/postgrest#2563]
Do you mean that if `order=clients.name,name,clients.rank` were used, then we should also consider to support `or=(clients.name.like.Chris*,and(name.eq.TopProject,clients.rank.gt.1)`?

It is indeed hard to read, and `or=(clients(name).like.Chris*,and(name.eq.TopProject,clients(rank).gt.1)` seems slightly better.

_Originally posted by @Iced-Sun in https://github.com/PostgREST/postgrest/issues/1414#issuecomment-986187388_
      
Also previously discussed on https://github.com/PostgREST/postgrest/discussions/2014

**Comment by steve-chavez** [comment]
Having the `client`, `clientinfo` and `contact` tables:
```
$ postgrest-with-postgresql-15 psql

postgres=# table test.clientinfo;
 id | clientid |      other
----+----------+------------------
  1 |        1 | 123 Main St
  2 |        2 | 456 South 3rd St
  3 |        3 | 789 Palm Tree Ln
(3 rows)

postgres=# table test.clientinfo;
 id | clientid |      other
----+----------+------------------
  1 |        1 | 123 Main St
  2 |        2 | 456 South 3rd St
  3 |        3 | 789 Palm Tree Ln
(3 rows)

postgres=# table test.contact;
 id |     name      | clientid
----+---------------+----------
  1 | Wally Walton  |        1
  2 | Wilma Wellers |        1
  3 | Tabby Targo   |        2
  4 | Bobby Bots    |        3
  5 | Bonnie Bits   |        3
  6 | Billy Boats   |        3
(6 rows)

```

I'd like to get the `client` rows where `clientinfo.name LIKE %Tabby% OR contact.other LIKE %Main%`.

Keeping the shape of our query, this could be done like:

```sql
  WITH pgrst_source AS (
    SELECT
      "test"."client".*
    FROM "test"."client"
    LEFT JOIN LATERAL (
      SELECT json_agg("client_clientinfo_1") AS "client_clientinfo_1"
      FROM (
        SELECT
          "clientinfo_1".*
        FROM "test"."clientinfo" AS "clientinfo_1"
        WHERE
        "clientinfo_1"."clientid" = "test"."client"."id" AND
        "clientinfo_1"."other" LIKE '%Main%'
      ) AS "client_clientinfo_1"
    ) AS "client_clientinfo_1" ON TRUE
    LEFT JOIN LATERAL (
      SELECT json_agg("client_contact_1") AS "client_contact_1"
      FROM (
        SELECT "contact_1".*
        FROM "test"."contact" AS "contact_1"
        WHERE
        "contact_1"."clientid" = "test"."client"."id" AND
        "contact_1"."name" LIKE '%Tabby%'
      ) AS "client_contact_1"
    ) AS "client_contact_1" ON TRUE
    WHERE
      "client_clientinfo_1" IS NOT NULL OR
      "client_contact_1" IS NOT NULL
  )
  SELECT coalesce(json_agg(_postgrest_t), '[]')::character varying AS body
  FROM ( SELECT * FROM pgrst_source ) _postgrest_t;

             body
------------------------------
 [{"id":1,"name":"Walmart"}, +
  {"id":2,"name":"Target"}]
```

Which could be expressed in the URL as:

```http
GET /client?select=*,clientinfo(),contact()&clientinfo.name=like.*Tabby*&contact.other=like.*Main*&or=(clientinfo.not.is.null,contact.not.is.null)
```

So the good news is that basically we only need to support doing `embed=not.is.null`, which is also a replacement for `!inner`.

**Comment by steve-chavez** [comment]
Reopening. It's a bit inconvenient(also unclear) to type the `not null` condition on the resource.

Related conditions would be more clear as per the original idea [above](https://github.com/PostgREST/postgrest/issues/2563#issue-1441072875).

So I think we could translate this:

```http
GET /projects?select=name,clients(*)&clients(id)=eq.1
```

To:

```http
GET /projects?select=name,clients(*)&clients.id=eq.1&clients=not.is.null
```

---

`or` would also be clearer as

```http
GET /client?select=*,clientinfo(),contact()&or=(clientinfo(other).like.*Main*,contact(name).like.*Tabby*)
```

Then translated to

```http
GET /client?select=*,clientinfo(),contact()&clientinfo.other=like.*Main*&contact.name=like.*Tabby*&or=(clientinfo.not.is.null,contact.not.is.null)
```

**Comment by thomasdohyunkim** [comment]
Hi there. Just wanted to check if there is any support for this in the works or any available workarounds to chain multiple `OR` filters on sub tables? I came here from [this stackoverflow questions](https://stackoverflow.com/questions/70699495/supabase-filter-data-with-join-and-logical-or).

**Comment by laurenceisla** [comment]
@thomasdohyunkim AFAIK this is already implemented,  but there's an issue with the clarity of the queries as Steve mentioned [above](#issuecomment-1361775232).  Here's an example on [how to use it](#issuecomment-1336573928).

The [latest pre-release](https://github.com/PostgREST/postgrest/releases/tag/v10.2.0.20230209) has this feature available.

**Comment by Etwenn** [comment]
Hello, i need to do almost the same thing, i.e : 
table1 : fieldA, fieldB + Foreign key to table2
table2 : fieldD, fieldE

i want to do :

GET /table1?select=*,table2(*)&or=(table2.fieldD.eq.thing,table2.fieldE.eq.thing

you would tell me to use this sentence i suppose :
GET /table1?select=*,table2(*)&table2.or=(fieldD.eq.thing,fieldE.eq.thing)

but sometime i need to do : 

GET /table1?select=*,table2(*)&or=(table2.fieldD.eq.thing,table2.fieldE.eq.thing,fieldA.eq.thing)

Here i'm stuck because i don't see how to use @steve-chavez workarround with multiple field of table2 + one field of table A

Do you have an idea ?

**Comment by laurenceisla** [comment]
> you would tell me to use this sentence i suppose :
`GET /table1?select=*,table2(*)&table2.or=(fieldD.eq.thing,fieldE.eq.thing)`

@Etwenn  Yes, and if you need to exclude the null embeddings )[top level filtering](https://postgrest.org/en/stable/references/api/resource_embedding.html#top-level-filtering)) you need to add `!inner` or `table2=not.is.null` to the query string. Both of these return the same result:

```
GET /table1?select=*,table2!inner(*)&table2.or=(fieldD.eq.thing,fieldE.eq.thing)

GET /table1?select=*,table2(*)&table2.or=(fieldD.eq.thing,fieldE.eq.thing)&table2=not.is.null
```

> but sometime i need to do : 
>
> `GET /table1?select=*,table2(*)&or=(table2.fieldD.eq.thing,table2.fieldE.eq.thing,fieldA.eq.thing)`

In this case, the filter needs to be in a separate query parameter and after that include the `table2=not.is.null` inside the `or` filter. It would be something like this:

```
GET /table1?select=*,table2(*)&table2.or=(fieldD.eq.thing,fieldE.eq.thing)&or=(fieldA.eq.thing,table2.not.is.null)
```

---

**NOTE:**  This is related to the issue #2800 that you opened. If the row has a column with a null value, then it won't be included in the response (when it should). But I don't see a workaround for this case.

**Comment by steve-chavez** [comment]
FYI, or conditions across embedded resources have been possible for a while now with https://postgrest.org/en/latest/references/api/resource_embedding.html#or-filtering-across-embedded-resources.

Though we can still improve the syntax to be more straightforward

**Comment by theshadowagent** [comment]
Any updates here? I need to do exactly the same thing as @Etwenn, and the complexity of the workaround proposed by @laurenceisla is driving me crazy. This is super non-intuitive for SDK users, and can be fixed only by guessing/googling.

**Comment by laurenceisla** [comment]
> `or` would also be clearer as
> 
> ```http
> GET /client?select=*,clientinfo(),contact()&or=(clientinfo(other).like.*Main*,contact(name).like.*Tabby*)
> ```
> 
> Then translated to
> 
> ```http
> GET /client?select=*,clientinfo(),contact()&clientinfo.other=like.*Main*&contact.name=like.*Tabby*&or=(clientinfo.not.is.null,contact.not.is.null)
> ```

@steve-chavez I agree with this proposal (at least for the `or` implementation), do you see if there's any new problem with it now? There are a lot of discussions/questions surrounding this and that solution could make it a lot clearer.

**Comment by steve-chavez** [comment]
> do you see if there's any new problem with it now?

@laurenceisla No, that proposal should still be good. And yes, everything would be clearer - even non-or cases, specifying `!inner` alone is already complicated.

**Comment by steve-chavez** [comment]
It would be a great DX improvement being able to do inner join conditions like:

```http
GET /projects?select=name,clients(*)&clients(id)=eq.1
```

(Instead of adding `!inner` or `clients=not.is.null`)

And regular left join conditions as always:

```http
GET /projects?select=name,clients(*)&clients.id=eq.1
```

And even greater that we can do this while maintaining backwards compat.

**Comment by steve-chavez** [comment]
> However, in this example, it seems like it will only ever be possible to put exactly one column in the (..) for the filter. 

Yes, only one column. This syntax is already used by the top-level ordering: https://docs.postgrest.org/en/v12/references/api/resource_embedding.html#top-level-ordering. 

```http
GET /films?order=directors(last_name).desc"
```

Given that we already have that, it's a natural extension of the syntax to support the same for filters. Top-level filters would be much clearer that way.

> What would the result of this be?
> GET /projects?select=name,clients(*)&clients(id)=eq.1&clients.name=eq.Doe
> Inner join or left join?

Not 100%, but IIRC `clients(id)=eq.1` doesn't actually do `INNER` and instead just adds a conditional to the LEFT LATERAL JOIN. So the above should work. Maybe @laurenceisla can clarify.

It's possible that we might have to clear docs and forget about the whole `inner` and `left` concepts though, and instead think about top-level and same level filters. top-level is same a "hoisting" concept which is used in aggs so for consistency this seems good.

**Comment by laurenceisla** [comment]
> This syntax is already used by the top-level ordering:

Yes, I agree here, I don't see much problem with the syntax personally.

> Apart from the syntax, there is one more problem, imho. Users would learn, that `clients(id)` means "inner join", and `clients.id` means "left join".

I wasn't really sure with the cases outside `or` across embeddings so I didn't put much thought into it. Now...

> What would the result of this be?
>
> ```http
> GET /projects?select=name,clients(*)&clients(id)=eq.1&clients.name=eq.Doe
> ```
 
Once a `clients(id)=eq.1`  is selected,  then a `clients=not.is.null` would be added. So yeah, the `clients.name=eq.Doe` would be applied to a `not null` client, making this example the same as:

```http
GET /projects?select=name,clients(*)&clients.id=eq.1&clients.name=eq.Doe&clients=not.is.null
```

I understand that it could get confusing, though. We'd need to document this as "once a related filter (`embed(columnA)=eq.val`) is present, the other filters (`embed.columnX=eq.val`) will also affect the top relationship like an inner join".

The `or` case has a similar issue, but the filters would be added outside the `or=(...)`. For example, if we add an extra filter for the`clientinfo`, say `phone_number`, then a query like this:

```bash
curl --get 'localhost:3000/client' \
  -d 'select=*,clientinfo(),contact()' \
  -d 'clientinfo.phone_number=like.999*' \
  -d 'or=(clientinfo(other).like.*Main*,contact(name).like.*Tabby*)'
```

Would be translated to:

```bash
curl --get 'localhost:3000/client' \
  -d 'select=*,clientinfo(),contact()' \
  -d 'clientinfo.phone_number=like.999*' \
  -d 'clientinfo.other=like.*Main*' \
  -d 'contact.name=like.*Tabby*' \
  -d 'or=(clientinfo.not.is.null,contact.not.is.null)'
```

This is the same, once the related `or` filter is present then any filter added to that relationship will affect the top query like an inner join. In this case, making the `or=(clientinfo.not.is.null,contact.not.is.null)` explicit is cumbersome for users as many issues/disussions can attest.
